### PR TITLE
statistics: add bucket structure.

### PR DESCRIPTION
**Pull request by hanfei1991**
add Buctket structure and adjust the way of estimating.

@shenli @coocood @zimulala @lamxTyler PTAL

### CODE REVIEW

File: statistics/builder.go [hunk 1]
```diff
@@ -263,51 +261,56 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
...
 		}
 		if cmp == 0 {
 			// The new item has the same value as current bucket value, to ensure that
 			// a same value only stored in a single bucket, we do not increase bucketIdx even if it exceeds
 			// valuesPerBucket.
-			col.Numbers[bucketIdx] = i * sampleFactor
-			col.Repeats[bucketIdx] += sampleFactor
+			col.Buckets[bucketIdx].Count = (i + 1) * sampleFactor
+			col.Buckets[bucketIdx].Repeats += sampleFactor
 		} else if i*sampleFactor-lastNumber <= valuesPerBucket {
```
Comment (coocood) [inline statistics/builder.go]: This should be  `(i+1)*sampleFactor-lastNumber)`.

File: statistics/column.go [hunk 2]
```diff
@@ -239,19 +227,17 @@ func (c *Column) search(sc *variable.StatementContext, target types.Datum) (inde
 func (c *Column) mergeBuckets(bucketIdx int64) {
 	curBuck := 0
 	for i := int64(0); i+1 <= bucketIdx; i += 2 {
-		c.Numbers[curBuck] = c.Numbers[i+1]
-		c.Values[curBuck] = c.Values[i+1]
-		c.Repeats[curBuck] = c.Repeats[i+1]
+		c.Buckets[curBuck].Count = c.Buckets[i+1].Count
+		c.Buckets[curBuck].Value = c.Buckets[i+1].Value
+		c.Buckets[curBuck].Repeats = c.Buckets[i+1].Repeats
```
Comment (lamxTyler) [inline statistics/column.go]: Why not directly copy the whole bucket?

File: statistics/column.go [hunk 3]
```diff
@@ -99,99 +100,85 @@ func colStatsFromStorage(ctx context.Context, tableID int64, colID int64, tp *ty
...
-	nextNumber := int64(0)
-	if index < len(c.Numbers)-1 {
-		nextNumber = c.Numbers[index+1]
-	}
-	greaterThanBucketValueCount := number - c.Repeats[index]
-	if match {
-		return greaterThanBucketValueCount, nil
-	}
-	return (nextNumber + greaterThanBucketValueCount) / 2, nil
+	return c.totalRowCount() - lessCount, nil
```
Comment (lamxTyler) [inline statistics/column.go]: Greater count should be total count minus less or equal count.

File: statistics/builder.go [hunk 4]
```diff
@@ -263,51 +261,56 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
...
+			// TODO: Making sampleFactor as float may be better.
 			// The bucket still have room to store a new item, update the bucket.
-			col.Numbers[bucketIdx] = i * sampleFactor
-			col.Values[bucketIdx] = samples[i]
-			col.Repeats[bucketIdx] = 0
+			col.Buckets[bucketIdx].Count = (i + 1) * sampleFactor
+			col.Buckets[bucketIdx].Value = samples[i]
+			col.Buckets[bucketIdx].Repeats = sampleFactor
```
Comment (coocood) [inline statistics/builder.go]: Use NDV is more accurate.

File: statistics/column.go [hunk 5]
```diff
@@ -239,19 +227,17 @@ func (c *Column) search(sc *variable.StatementContext, target types.Datum) (inde
 func (c *Column) mergeBuckets(bucketIdx int64) {
 	curBuck := 0
 	for i := int64(0); i+1 <= bucketIdx; i += 2 {
-		c.Numbers[curBuck] = c.Numbers[i+1]
-		c.Values[curBuck] = c.Values[i+1]
-		c.Repeats[curBuck] = c.Repeats[i+1]
+		c.Buckets[curBuck].Count = c.Buckets[i+1].Count
```
Comment (coocood) [inline statistics/column.go]:
```
c.Bucket[curBuck] = Bucket{
     ...
}
```

File: statistics/builder.go [hunk 6]
```diff
@@ -263,51 +261,56 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
...
 		} else {
 			// The bucket is full, store the item in the next bucket.
-			lastNumber = col.Numbers[bucketIdx]
```
Comment (lamxTyler) [inline statistics/builder.go]: Why delete this?

Comment (hanfei1991) [inline statistics/builder.go]: it's use less, count is calculate by (i+1) * sampleFactor

Comment (lamxTyler) [inline statistics/builder.go]: If you delete this, then when you test if this bucket is full, it will always false, which will lead to too many buckets.

Comment (hanfei1991) [inline statistics/builder.go]: Well, you are right

Comment (lamxTyler) [inline statistics/builder.go]: I think we need a test to make sure that the number of buckets do not exceed the predefined bucket count.

File: statistics/builder.go [hunk 7]
```diff
@@ -252,62 +253,75 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
...
-		Values:  make([]types.Datum, 1, bucketCount),
-		Repeats: make([]int64, 1, bucketCount),
+		Buckets: make([]bucket, 1, bucketCount),
 	}
 	valuesPerBucket := t.Count/bucketCount + 1
 
 	// As we use samples to build the histogram, the bucket number and repeat should multiply a factor.
 	sampleFactor := t.Count / int64(len(samples))
+	ndvFactor := t.Count / ndv
+	log.Warnf("sample %d ndv %d ndvFact %d", sampleFactor, ndv, ndvFactor)
```
Comment (lamxTyler) [inline statistics/builder.go]: Is this log necessary?

**Comment by lamxTyler** [comment]
LGTM

File: statistics/column.go [hunk 8]
```diff
@@ -31,19 +31,22 @@ type Column struct {
...
-	// TODO: We could have make a bucket struct contains number, value and repeats.
-	Numbers []int64
-	Values  []types.Datum
-	Repeats []int64
+	Buckets []bucket
+}
+
+// bucket is an element of histogram.
+//
+// A bucket number is the number of items stored in all previous buckets and the current bucket.
```
Comment (zimulala) [inline statistics/column.go]: Update the comment.

File: statistics/builder.go [hunk 9]
```diff
@@ -252,62 +252,74 @@ func (t *Table) buildColumn(sc *variable.StatementContext, offset int, ndv int64
...
 			return errors.Trace(err)
 		}
 		if cmp == 0 {
 			// The new item has the same value as current bucket value, to ensure that
 			// a same value only stored in a single bucket, we do not increase bucketIdx even if it exceeds
 			// valuesPerBucket.
-			col.Numbers[bucketIdx] = i * sampleFactor
-			col.Repeats[bucketIdx] += sampleFactor
-		} else if i*sampleFactor-lastNumber <= valuesPerBucket {
+			col.Buckets[bucketIdx].Count = (i + 1) * sampleFactor
```
Comment (zimulala) [inline statistics/builder.go]: Define a variable for `i+1`. Many places use it.

**Comment by zimulala** [comment]
LGTM

**Pull request merged**